File: WinForms/CS/ChartsMainDemo/Modules/ChartViews/RadarViews.cs
using System;
using System.Collections.Generic;
using DevExpress.Utils;
using DevExpress.XtraBars.Docking2010.Views;
using DevExpress.XtraBars.Docking2010.Views.Tabbed;
using DevExpress.XtraEditors;
using DevExpress.XtraLayout;
using DevExpress.XtraLayout.Utils;

namespace DevExpress.XtraCharts.Demos {
    public partial class RadarViewsDemo : ChartDemoModuleMarkerViews {
        bool IsScatterLineSeriesActive {
            get { return ActiveSeries.View is ScatterRadarLineSeriesView; }
        }

        protected override LayoutControl OptionsLayoutControl {
            get { return layoutControl; }
        }
        protected override CheckEdit LabelVisibleCheckEdit {
            get { return checkEditLabelVisible; }
        }
        protected override ComboBoxEdit MarkerKindComboBox {
            get { return comboBoxEditMarkerKind; }
        }
        protected override SpinEdit MarkerSizeSpinEdit {
            get { return spinEditMarkerSize; }
        }
        protected override CheckEdit MarkerVisibleCheckEdit {
            get { return checkEditMarkerVisible; }
        }
        protected override TabbedView TabbedView {
            get { return currentTabbedView; }
        }

        internal override List<ChartControl> ChartControls {
            get {
                return new List<ChartControl>() {
                    chartPoint,
                    chartLine,
                    chartArea,
                    chartRangeArea,
                    chartScatterLine
                };
            }
        }

        public RadarViewsDemo() {
            InitializeComponent();
        }

        void tabbedView_DocumentActivated(object sender, DocumentEventArgs e) {
            OnDocumentActivated(e.Document);
            if(IsScatterLineSeriesActive)
                comboBoxEditFunction.SelectedIndex = 0;
        }
        void checkEditLabelVisible_CheckedChanged(object sender, EventArgs e) {
            SetLabelVisibility();
        }
        void checkEditMarkerVisible_CheckedChanged(object sender, EventArgs e) {
            SetMarkerVisibility(layoutControlItemMarkerKind, layoutControlItemMarkerSize);
        }
        void comboBoxEditTextDirection_SelectedIndexChanged(object sender, EventArgs e) {
            RadarAxisXLabelTextDirection textDirection = ((DirectionComboBoxItem)comboBoxEditTextDirection.EditValue).Direction;
            RadarDiagram diagram = ChartControl.Diagram as RadarDiagram;
            if(diagram != null)
                diagram.AxisX.Label.TextDirection = textDirection;
        }
        void comboBoxEditDiagramStyle_SelectedIndexChanged(object sender, EventArgs e) {
            RadarDiagram diagram = ChartControl.Diagram as RadarDiagram;
            if(diagram != null)
                diagram.DrawingStyle = (RadarDiagramDrawingStyle)comboBoxDiagramStyle.SelectedIndex;
        }
        void comboBoxEditFunction_SelectedIndexChanged(object sender, EventArgs e) {
            Series series = ActiveSeries as Series;
            if(series == null)
                return;
            series.Points.Clear();
            RadianScatterFunctionCalculator functionCalculator = new RadianScatterFunctionCalculator();
            SeriesPoint[] points = functionCalculator.GenerateScatterFunctionPoints(comboBoxEditFunction.SelectedIndex);
            series.Points.AddRange(points);
            ChartControl.Animate();
        }
        void comboBoxEditRotationDirection_SelectedIndexChanged(object sender, EventArgs e) {
            RadarDiagram diagram = ChartControl.Diagram as RadarDiagram;
            if(diagram != null)
                diagram.RotationDirection = (RadarDiagramRotationDirection)comboBoxEditRotationDirection.SelectedIndex;
        }
        void spinEditStartAngle_EditValueChanged(object sender, EventArgs e) {
            RadarDiagram diagram = ChartControl.Diagram as RadarDiagram;
            if(diagram != null)
                diagram.StartAngleInDegrees = (int)spinEditStartAngle.Value;
        }
        void spinEditAxisLabelAngle_EditValueChanged(object sender, EventArgs e) {
            RadarDiagram diagram = ChartControl.Diagram as RadarDiagram;
            if(diagram != null)
                diagram.AxisY.Label.Angle = (int)spinEditAxisLabelAngle.Value;
        }

        protected override void InitControls() {
            base.InitControls();
            Array directions = Enum.GetValues(typeof(RadarAxisXLabelTextDirection));
            foreach(RadarAxisXLabelTextDirection direction in directions)
                comboBoxEditTextDirection.Properties.Items.Add(new DirectionComboBoxItem(direction));
        }
        protected override void OnLoad(EventArgs e) {
            base.OnLoad(e);
            documentManager.View.Controller.Activate(documentPoint);
        }
        protected override void UpdateControlsCore() {
            base.UpdateControlsCore();
            layoutControlItemTextDirection.Visibility = IsScatterLineSeriesActive ? LayoutVisibility.Never : LayoutVisibility.Always;
            layoutControlItemFunction.Visibility = IsScatterLineSeriesActive ? LayoutVisibility.Always : LayoutVisibility.Never;
            layoutControlGroupGeneral.Visibility = IsScatterLineSeriesActive ? LayoutVisibility.Always : LayoutVisibility.Never;
            checkEditLabelVisible.Checked = ActiveSeries.LabelsVisibility == DefaultBoolean.True;
            RadarDiagram diagram = ChartControl.Diagram as RadarDiagram;
            if(diagram != null) {
                comboBoxEditTextDirection.EditValue = new DirectionComboBoxItem(diagram.AxisX.Label.TextDirection);
                comboBoxDiagramStyle.SelectedIndex = (int)diagram.DrawingStyle;
                comboBoxEditRotationDirection.SelectedIndex = (int)diagram.RotationDirection;
                spinEditStartAngle.EditValue = diagram.StartAngleInDegrees;
                spinEditAxisLabelAngle.EditValue = diagram.AxisY.Label.Angle;
            }
            layoutControlItemMarkerVisible.Visibility = UpdateRadarMarkerOptions();
        }
        protected override bool SetRangeMarkerVisibility(SeriesViewBase view, DefaultBoolean visibility) {
            RadarRangeAreaSeriesView rangeAreaView = view as RadarRangeAreaSeriesView;
            if(rangeAreaView == null)
                return false;
            rangeAreaView.Marker1Visibility = visibility;
            rangeAreaView.Marker2Visibility = visibility;
            return true;
        }
        protected override bool SetRangeMarkerSize(SeriesViewBase view, int size) {
            RadarRangeAreaSeriesView rangeAreaView = view as RadarRangeAreaSeriesView;
            if(rangeAreaView == null)
                return false;
            rangeAreaView.Marker1.Size = size;
            rangeAreaView.Marker2.Size = size;
            return true;
        }
        protected override bool SetRangeMarkerKind(SeriesViewBase view, MarkerKind kind, int starCount) {
            RadarRangeAreaSeriesView rangeAreaView = view as RadarRangeAreaSeriesView;
            if(rangeAreaView == null)
                return false;
            rangeAreaView.Marker1.Kind = kind;
            rangeAreaView.Marker1.StarPointCount = starCount;
            rangeAreaView.Marker2.Kind = kind;
            rangeAreaView.Marker2.StarPointCount = starCount;
            return true;
        }
    }


    class DirectionComboBoxItem {
        public string Text {
            get;
            private set;
        }
        public RadarAxisXLabelTextDirection Direction {
            get;
            private set;
        }

        public DirectionComboBoxItem(RadarAxisXLabelTextDirection textDirection) {
            Direction = textDirection;
            Text = Utils.SplitByWords(textDirection.ToString()).Replace("To ", "to ");
        }

        public override bool Equals(object obj) {
            if(obj == null)
                return false;
            DirectionComboBoxItem directionComboBoxItem = obj as DirectionComboBoxItem;
            if(directionComboBoxItem == null)
                return false;
            return (Direction == directionComboBoxItem.Direction);
        }
        public override int GetHashCode() {
            return Direction.GetHashCode();
        }
        public override string ToString() {
            return Text;
        }
    }

}

File: WinForms/VB/ChartsMainDemo/Modules/ChartViews/RadarViews.vb
Imports System
Imports System.Collections.Generic
Imports DevExpress.Utils
Imports DevExpress.XtraBars.Docking2010.Views
Imports DevExpress.XtraBars.Docking2010.Views.Tabbed
Imports DevExpress.XtraEditors
Imports DevExpress.XtraLayout
Imports DevExpress.XtraLayout.Utils

Namespace DevExpress.XtraCharts.Demos

    Public Partial Class RadarViewsDemo
        Inherits ChartDemoModuleMarkerViews

        Private ReadOnly Property IsScatterLineSeriesActive As Boolean
            Get
                Return TypeOf ActiveSeries.View Is ScatterRadarLineSeriesView
            End Get
        End Property

        Protected Overrides ReadOnly Property OptionsLayoutControl As LayoutControl
            Get
                Return layoutControl
            End Get
        End Property

        Protected Overrides ReadOnly Property LabelVisibleCheckEdit As CheckEdit
            Get
                Return checkEditLabelVisible
            End Get
        End Property

        Protected Overrides ReadOnly Property MarkerKindComboBox As ComboBoxEdit
            Get
                Return comboBoxEditMarkerKind
            End Get
        End Property

        Protected Overrides ReadOnly Property MarkerSizeSpinEdit As SpinEdit
            Get
                Return spinEditMarkerSize
            End Get
        End Property

        Protected Overrides ReadOnly Property MarkerVisibleCheckEdit As CheckEdit
            Get
                Return checkEditMarkerVisible
            End Get
        End Property

        Protected Overrides ReadOnly Property TabbedView As TabbedView
            Get
                Return currentTabbedView
            End Get
        End Property

        Friend Overrides ReadOnly Property ChartControls As List(Of ChartControl)
            Get
                Return New List(Of ChartControl)() From {chartPoint, chartLine, chartArea, chartRangeArea, chartScatterLine}
            End Get
        End Property

        Public Sub New()
            InitializeComponent()
        End Sub

        Private Sub tabbedView_DocumentActivated(ByVal sender As Object, ByVal e As DocumentEventArgs)
            OnDocumentActivated(e.Document)
            If IsScatterLineSeriesActive Then comboBoxEditFunction.SelectedIndex = 0
        End Sub

        Private Sub checkEditLabelVisible_CheckedChanged(ByVal sender As Object, ByVal e As EventArgs)
            SetLabelVisibility()
        End Sub

        Private Sub checkEditMarkerVisible_CheckedChanged(ByVal sender As Object, ByVal e As EventArgs)
            SetMarkerVisibility(layoutControlItemMarkerKind, layoutControlItemMarkerSize)
        End Sub

        Private Sub comboBoxEditTextDirection_SelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs)
            Dim textDirection As RadarAxisXLabelTextDirection = CType(comboBoxEditTextDirection.EditValue, DirectionComboBoxItem).Direction
            Dim diagram As RadarDiagram = TryCast(ChartControl.Diagram, RadarDiagram)
            If diagram IsNot Nothing Then diagram.AxisX.Label.TextDirection = textDirection
        End Sub

        Private Sub comboBoxEditDiagramStyle_SelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs)
            Dim diagram As RadarDiagram = TryCast(ChartControl.Diagram, RadarDiagram)
            If diagram IsNot Nothing Then diagram.DrawingStyle = CType(comboBoxDiagramStyle.SelectedIndex, RadarDiagramDrawingStyle)
        End Sub

        Private Sub comboBoxEditFunction_SelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs)
            Dim series As Series = TryCast(ActiveSeries, Series)
            If series Is Nothing Then Return
            series.Points.Clear()
            Dim functionCalculator As RadianScatterFunctionCalculator = New RadianScatterFunctionCalculator()
            Dim points As SeriesPoint() = functionCalculator.GenerateScatterFunctionPoints(comboBoxEditFunction.SelectedIndex)
            series.Points.AddRange(points)
            ChartControl.Animate()
        End Sub

        Private Sub comboBoxEditRotationDirection_SelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs)
            Dim diagram As RadarDiagram = TryCast(ChartControl.Diagram, RadarDiagram)
            If diagram IsNot Nothing Then diagram.RotationDirection = CType(comboBoxEditRotationDirection.SelectedIndex, RadarDiagramRotationDirection)
        End Sub

        Private Sub spinEditStartAngle_EditValueChanged(ByVal sender As Object, ByVal e As EventArgs)
            Dim diagram As RadarDiagram = TryCast(ChartControl.Diagram, RadarDiagram)
            If diagram IsNot Nothing Then diagram.StartAngleInDegrees = CInt(spinEditStartAngle.Value)
        End Sub

        Private Sub spinEditAxisLabelAngle_EditValueChanged(ByVal sender As Object, ByVal e As EventArgs)
            Dim diagram As RadarDiagram = TryCast(ChartControl.Diagram, RadarDiagram)
            If diagram IsNot Nothing Then diagram.AxisY.Label.Angle = CInt(spinEditAxisLabelAngle.Value)
        End Sub

        Protected Overrides Sub InitControls()
            MyBase.InitControls()
            Dim directions As Array = [Enum].GetValues(GetType(RadarAxisXLabelTextDirection))
            For Each direction As RadarAxisXLabelTextDirection In directions
                comboBoxEditTextDirection.Properties.Items.Add(New DirectionComboBoxItem(direction))
            Next
        End Sub

        Protected Overrides Sub OnLoad(ByVal e As EventArgs)
            MyBase.OnLoad(e)
            documentManager.View.Controller.Activate(documentPoint)
        End Sub

        Protected Overrides Sub UpdateControlsCore()
            MyBase.UpdateControlsCore()
            layoutControlItemTextDirection.Visibility = If(IsScatterLineSeriesActive, LayoutVisibility.Never, LayoutVisibility.Always)
            layoutControlItemFunction.Visibility = If(IsScatterLineSeriesActive, LayoutVisibility.Always, LayoutVisibility.Never)
            layoutControlGroupGeneral.Visibility = If(IsScatterLineSeriesActive, LayoutVisibility.Always, LayoutVisibility.Never)
            checkEditLabelVisible.Checked = ActiveSeries.LabelsVisibility = DefaultBoolean.True
            Dim diagram As RadarDiagram = TryCast(ChartControl.Diagram, RadarDiagram)
            If diagram IsNot Nothing Then
                comboBoxEditTextDirection.EditValue = New DirectionComboBoxItem(diagram.AxisX.Label.TextDirection)
                comboBoxDiagramStyle.SelectedIndex = CInt(diagram.DrawingStyle)
                comboBoxEditRotationDirection.SelectedIndex = CInt(diagram.RotationDirection)
                spinEditStartAngle.EditValue = diagram.StartAngleInDegrees
                spinEditAxisLabelAngle.EditValue = diagram.AxisY.Label.Angle
            End If

            layoutControlItemMarkerVisible.Visibility = UpdateRadarMarkerOptions()
        End Sub

        Protected Overrides Function SetRangeMarkerVisibility(ByVal view As SeriesViewBase, ByVal visibility As DefaultBoolean) As Boolean
            Dim rangeAreaView As RadarRangeAreaSeriesView = TryCast(view, RadarRangeAreaSeriesView)
            If rangeAreaView Is Nothing Then Return False
            rangeAreaView.Marker1Visibility = visibility
            rangeAreaView.Marker2Visibility = visibility
            Return True
        End Function

        Protected Overrides Function SetRangeMarkerSize(ByVal view As SeriesViewBase, ByVal size As Integer) As Boolean
            Dim rangeAreaView As RadarRangeAreaSeriesView = TryCast(view, RadarRangeAreaSeriesView)
            If rangeAreaView Is Nothing Then Return False
            rangeAreaView.Marker1.Size = size
            rangeAreaView.Marker2.Size = size
            Return True
        End Function

        Protected Overrides Function SetRangeMarkerKind(ByVal view As SeriesViewBase, ByVal kind As MarkerKind, ByVal starCount As Integer) As Boolean
            Dim rangeAreaView As RadarRangeAreaSeriesView = TryCast(view, RadarRangeAreaSeriesView)
            If rangeAreaView Is Nothing Then Return False
            rangeAreaView.Marker1.Kind = kind
            rangeAreaView.Marker1.StarPointCount = starCount
            rangeAreaView.Marker2.Kind = kind
            rangeAreaView.Marker2.StarPointCount = starCount
            Return True
        End Function
    End Class

    Friend Class DirectionComboBoxItem

        Private _Text As String, _Direction As RadarAxisXLabelTextDirection

        Public Property Text As String
            Get
                Return _Text
            End Get

            Private Set(ByVal value As String)
                _Text = value
            End Set
        End Property

        Public Property Direction As RadarAxisXLabelTextDirection
            Get
                Return _Direction
            End Get

            Private Set(ByVal value As RadarAxisXLabelTextDirection)
                _Direction = value
            End Set
        End Property

        Public Sub New(ByVal textDirection As RadarAxisXLabelTextDirection)
            Direction = textDirection
            Text = SplitByWords(textDirection.ToString()).Replace("To ", "to ")
        End Sub

        Public Overrides Function Equals(ByVal obj As Object) As Boolean
            If obj Is Nothing Then Return False
            Dim directionComboBoxItem As DirectionComboBoxItem = TryCast(obj, DirectionComboBoxItem)
            If directionComboBoxItem Is Nothing Then Return False
            Return Direction = directionComboBoxItem.Direction
        End Function

        Public Overrides Function GetHashCode() As Integer
            Return Direction.GetHashCode()
        End Function

        Public Overrides Function ToString() As String
            Return Text
        End Function
    End Class
End Namespace
